### PR TITLE
feat : added state update worker

### PR DESCRIPTION
**Pull request by ocdbytes**

### CODE REVIEW

File: crates/orchestrator/src/database/mongodb/mod.rs [hunk 1]
```diff
@@ -244,4 +244,49 @@ impl Database for MongoDb {
...
+    async fn get_last_successful_job_by_type(&self, job_type: JobType) -> Result<Option<JobItem>> {
+        let filter = doc! {
+            "job_type": bson::to_bson(&job_type)?,
+            "job_status": bson::to_bson(&JobStatus::Completed)?
+        };
+        let find_options = FindOneOptions::builder().sort(doc! { "internal_id": -1 }).build();
+
+        Ok(self
+            .get_job_collection()
+            .find_one(filter, find_options)
```
Comment (apoorvsadana) [inline crates/orchestrator/src/database/mongodb/mod.rs:257]: we should index collections

File: crates/orchestrator/src/workers/update_state.rs [hunk 2]
```diff
@@ -8,10 +11,46 @@ pub struct UpdateStateWorker;
...
+                    .get_completed_jobs_after_internal_id_by_job_type(
+                        JobType::ProofCreation,
+                        latest_successful_job_internal_id,
+                    )
+                    .await?;
+
+                for job in successful_proving_jobs {
+                    let existing_job = config
+                        .database()
+                        .get_job_by_internal_id_and_type(&job.internal_id, &JobType::StateTransition)
```
Comment (apoorvsadana) [inline crates/orchestrator/src/workers/update_state.rs:36]: we can remove this

File: crates/orchestrator/src/workers/update_state.rs [hunk 3]
```diff
@@ -8,10 +11,46 @@ pub struct UpdateStateWorker;
...
+        let config = config().await;
+        let latest_successful_job = config.database().get_last_successful_job_by_type(JobType::StateTransition).await?;
+
+        match latest_successful_job {
+            Some(job) => {
+                let latest_successful_job_internal_id = job.internal_id;
+
+                let successful_proving_jobs = config
+                    .database()
+                    .get_completed_jobs_after_internal_id_by_job_type(
```
Comment (apoorvsadana) [inline crates/orchestrator/src/workers/update_state.rs:27]: let's move completed to an argument itself

File: crates/orchestrator/src/tests/workers/utils/mod.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        internal_id: id.clone(),
+        job_type: JobType::SnosRun,
+        status: JobStatus::Created,
+        external_id: ExternalId::Number(0),
+        metadata: HashMap::new(),
+        version: 0,
+    }
+}
+
+pub fn get_job_by_mock_id_vector(
```
Comment (apoorvsadana) [inline crates/orchestrator/src/tests/workers/utils/mod.rs:20]: let's add a comment on this

File: crates/orchestrator/src/tests/workers/update_state/mod.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            .returning(move |_, _| {
+                Ok(get_job_by_mock_id_vector(
+                    JobType::ProofCreation,
+                    JobStatus::Completed,
+                    number_of_processed_jobs as u64,
+                    2,
+                ))
+            });
+
+        // mocking getting of the jobs
```
Comment (apoorvsadana) [inline crates/orchestrator/src/tests/workers/update_state/mod.rs:59]: let's add a comment on where this check happens

File: crates/orchestrator/src/database/mongodb/mod.rs (hunk 1 above)
Comment (ocdbytes) [reply crates/orchestrator/src/database/mongodb/mod.rs:257]: will create the issue for this.

File: lcov.info
Comment (apoorvsadana) [inline lcov.info:1]: does this file need to be in a .gitignore?

File: crates/orchestrator/src/tests/workers/update_state/mod.rs (hunk 5 above)
Comment (apoorvsadana) [reply crates/orchestrator/src/tests/workers/update_state/mod.rs:58]: I guess this is pending?

**Pull request merged**